domain/macaroon/service/rootkeys.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/bakery"
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/bakery/dbrootkeystore"
+	"github.com/juju/errors"
+
+	macaroonerrors "github.com/juju/juju/domain/macaroon/errors"
+)
+
+// RootKeyState describes the persistence layer for
+// macaroon root keys
+type RootKeyState interface {
+	// GetKey gets the key with a given id from state. If not key is found, a
+	// macaroonerrors.KeyNotFound error is returned.
+	GetKey(ctx context.Context, id []byte) (dbrootkeystore.RootKey, error)
+
+	// FindLatestKey returns the most recently created root key k following all
+	// the conditions:
+	//
+	// k.Created >= createdAfter
+	// k.Expires >= expiresAfter
+	// k.Expires <= expiresBefore
+	//
+	// If no such key was found, return a macaroonerrors.KeyNotFound error
+	FindLatestKey(ctx context.Context, createdAfter, expiresAfter, expiresBefore time.Time) (dbrootkeystore.RootKey, error)
+
+	// InsertKey inserts the given root key into state. If a key with matching
+	// id already exists, return a macaroonerrors.KeyAlreadyExists error.
+	InsertKey(ctx context.Context, key dbrootkeystore.RootKey) error
+}
```

Comment:
This is incorrect. The `dbrootkeystore.RootKey` should be normalised at the service layer and converted to an internal macroon database type. At no point should the state know about dbrootkeystore.